### PR TITLE
Faster min masters

### PR DESCRIPTION
**Pull request by mkuratczyk**
Comparison of the new implementation and old implementation (+ random and client-local for reference)
```
+----------------------------+-------+-----------------+--------------+----------------+------------+-------------+
|            test            | nodes |    placement    | import (sec) | reimport (sec) | stop (sec) | start (sec) |
+----------------------------+-------+-----------------+--------------+----------------+------------+-------------+
| 100K classic queues        |     1 | new min-masters |       250.02 |          22.64 |      65.55 |    1,094.31 |
| 100K classic queues        |     1 | old min-masters |     2,579.64 |       5,471.84 |      80.39 |    1,411.49 |
| 100K classic queues        |     3 | client-local    |       267.58 |          26.94 |      58.19 |      564.73 |
| 100K classic queues        |     3 | new min-masters |       267.84 |         153.62 |      29.38 |      181.46 |
| 100K classic queues        |     3 | random          |       232.78 |          45.27 |      17.89 |      117.81 |
| 100K classic queues        |     3 | old min-masters |     2,627.43 |       5,401.61 |      20.52 |      124.06 |
| 100K classic queues        |     5 | new min-masters |       225.50 |         127.69 |      21.16 |      114.05 |
| 100K classic queues        |     5 | old min-masters |     2,457.90 |       5,016.11 |      23.94 |      119.44 |
| 5K classic mirrored queues |     1 | new min-masters |         5.50 |           2.59 |       2.98 |        6.56 |
| 5K classic mirrored queues |     1 | old min-masters |        10.31 |          12.29 |       3.08 |        7.04 |
| 5K classic mirrored queues |     3 | client-local    |         9.41 |           8.99 |       6.82 |       23.10 |
| 5K classic mirrored queues |     3 | new min-masters |        13.06 |           7.29 |       3.41 |       26.61 |
| 5K classic mirrored queues |     5 | new min-masters |        12.91 |           5.81 |       2.74 |        9.93 |
| 5K classic mirrored queues |     5 | old min-masters |        17.77 |          17.41 |       2.93 |       10.28 |
+----------------------------+-------+-----------------+--------------+----------------+------------+-------------+
```

### CODE REVIEW

File: deps/rabbit/src/rabbit_queue_location_min_masters.erl [hunk 1]
```diff
@@ -6,65 +6,60 @@
...
-            {ok, MinNode};
-        false ->
-            undefined
+    Nodes0 = rabbit_nodes:all_running(),
+    Nodes = rabbit_maintenance:filter_out_drained_nodes_local_read(Nodes0),
+    case Nodes of
+        [] ->
+            {ok, node()};
+        _ ->
+            MastersPerNode0 = erpc:multicall(Nodes, ets, info, [queue_metrics, size], 10),
```
Comment (michaelklishin) [inline deps/rabbit/src/rabbit_queue_location_min_masters.erl:49]: 10ms is an incredibly low timeout, please change this to 15s which is our default in most places

**Review by michaelklishin — CHANGES_REQUESTED** [review]

**Comment by michaelklishin** [comment]
I have also noticed a test failure that I could not reproduce with runs on `master`:

```
queue_master_location_SUITE > cluster_size_3 > declare_policy_nodes
    #1. {error,
            {{assertEqual,
                 [{module,queue_master_location_SUITE},
                  {line,454},
                  {expression,"Rpc"},
                  {expected,{ok,'declare_policy_nodes-2@localhost'}},
                  {value,{ok,'declare_policy_nodes-3@localhost'}}]},
             [{queue_master_location_SUITE,verify_min_master,3,
                  [{file,"test/queue_master_location_SUITE.erl"},{line,454}]},
              {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1783}]},
              {test_server,run_test_case_eval1,6,
                  [{file,"test_server.erl"},{line,1292}]},
              {test_server,run_test_case_eval,9,
                  [{file,"test_server.erl"},{line,1224}]}]}}
```

This can be a test interference that's been uncovered.

**Comment by lukebakken** [comment]
Hmm I get this build error -

```
%   13| -include_lib("rabbitmq_ct_helpers/include/rabbit_assert.hrl").
%     |              ^

test/maintenance_mode_SUITE.erl:251:14: undefined macro 'awaitMatch/3'
%  251|             ?awaitMatch(
%     |              ^

make[1]: *** [../../erlang.mk:5581: /home/bakkenl/development/rabbitmq/rabbitmq-server/deps/rabbit/.erlang.mk/rabbit.last-testdir-build] Error 1
make: *** [../../erlang.mk:5593: test-build] Error 2
```

Running this command from `deps/rabbit`:

```
make ct-queue_master_location t=cluster_size_3:declare_policy_nodes
```

**Comment by mkuratczyk** [comment]
Current implementations has a flaw: it relies on metrics that are refreshed periodically. Therefore, it's not really "min-masters": if the user declare a few queues quickly (eg. during import), they will likely end up on a single node, because all those queues will get declared, before stats are refreshed. To finish this change, I think we should introduce a new metrics - number of queues on the node, so that we have an accurate view of the distribution. This should also be even faster than calculating the size of an ETS table (which is relatively fast, but I saw it slows down as the number of queues grow)

**Comment by mkuratczyk** [comment]
If we want something like this, we should just use the `balanced` approach as introduced in https://github.com/rabbitmq/rabbitmq-server/pull/4442. With enough queues, it makes no sense to count them precisely and we can internally switch to random distribution which will be much easier/faster, while providing similar disribution

**Pull request closed without merging**